src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -95,19 +120,59 @@ public Http2Connection(Http2ConnectionContext context)
 
         public void OnInputOrOutputCompleted()
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState != ConnectionState.Closed)
+                {
+                    _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.NO_ERROR);
+                    _connectionState = ConnectionState.Closed;
+                }
+            }
+
             _frameWriter.Complete();
         }
 
         public void Abort(ConnectionAbortedException ex)
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState != ConnectionState.Closed)
+                {
+                    _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.INTERNAL_ERROR);
+                    _connectionState = ConnectionState.Closed;
+                }
+            }
+
             _frameWriter.Abort(ex);
         }
 
         public void StopProcessingNextRequest()
+            => StopProcessingNextRequest(true);
+
+        public void StopProcessingNextRequest(bool sendGoAway = false)
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState == ConnectionState.Open)
+                {
+                    if (_streams.IsEmpty)
+                    {
+                        _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.NO_ERROR);
```

Comment:
Why not honor sendGoAway = false on this code path?

Reply:
Only the graceful GOAWAY (with the highest opened stream id set to maxvalue) is optional. The final GOAWAY we send before closing the connection is a must.

Reply:
so sendGoAway would be true in that case? For consistency in this method you should still check it.

Reply:
Not quite, we set sendgoaway to false when processing a GOAWAY from the client. In this case, we don't want to respond by sending a graceful GOAWAY if there are streams still open. However, if there were no streams open, we want to close the connection right away and send the final GOAWAY with the highest opened stream id regardless of whether sendgoaway was set to false.

Reply:
Then sendGoAway is misnamed and you should come up with something more specific.